test_formulae: improve result caching

carlocab opened the pull request:
Our current method of using artifacts as a cache breaks when a PR is
pushed to multiple times in quick succession.

This is because we use the SHA1 prior to the most recent push to look up
artifacts, but there will be no useable artifacts to find there after
pushing to a PR that has recently been pushed to.

We can improve this by looking for an even earlier SHA1 to try to find
useable artifacts from.

This is a companion to Homebrew/homebrew-core#132093.

Bo98 commented on lib/test_formulae.rb:
```diff
@@ -15,22 +15,22 @@ def initialize(tap:, git:, dry_run:, fail_fast:, verbose:)
-          before
-        end
+        event_path = cached_event_json
+        event_path ||= Pathname(ENV.fetch("GITHUB_EVENT_PATH"))
+        event_payload = JSON.parse(event_path.read)
 
-        @previous_sha
+        event_payload.fetch("before", nil)
```
So if we have the current event we take the previous (before) sha, but if we download a previous artifact don't we want the current sha from that payload? Or will that payload be the current run from an earlier job?

Bo98 commented on lib/test_formulae.rb:
```diff
@@ -97,33 +97,48 @@ def download_artifact_from_previous_run!(name)
         artifact_cache.cd do
           GitHub.download_artifact(download_url, run_id)
         end
+
+        return if wanted_artifact_name == name
+
+        download_artifact_from_previous_run!(name)
```
This doesn't seem safe. We could get an event payload artifact from a different run from the bottles.

I feel like it's just going to be safer to just bundle the event payload (or a subset of it) in with the bottle artifact.

Bo98 commented on lib/test_formulae.rb:
```diff
@@ -97,33 +97,48 @@ def download_artifact_from_previous_run!(name)
 
-        wanted_artifact = response.fetch("artifacts").find { |artifact| artifact.fetch("name") == name }
+        artifacts = response.fetch("artifacts")
+        wanted_artifact = artifacts.find { |artifact| artifact.fetch("name") == name }
+        wanted_artifact ||= artifacts.find { |artifact| artifact.fetch("name") == "event_payload" }
         return if wanted_artifact.blank?
 
-        ohai "Downloading #{name} from #{previous_github_sha}"
+        wanted_artifact_name = wanted_artifact.fetch("name")
+        cached_event_json&.unlink if wanted_artifact_name == "event_payload"
+
+        ohai "Downloading #{wanted_artifact_name} from #{sha}"
         download_url = wanted_artifact.fetch("archive_download_url")
 
         artifact_cache.mkpath
         artifact_cache.cd do
           GitHub.download_artifact(download_url, run_id)
         end
+
+        return if wanted_artifact_name == name
+
+        download_artifact_from_previous_run!(name)
```
Honestly some comments here with the intended flow could be useful. I was confused at first as I thought this would have downloaded a bottle from a different run from the event payload.

carlocab replied:
Yup, that's exactly how this works. Will add some comments.

carlocab replied:
Added in 8b1655066fdc15b91dd2a203a87bea3b5944c1a9.

Bo98 commented:
Actually I have one problem here: what if the PR author tampers with the event_payload upload with a workflow edit? This seems possible.

Bo98 commented on lib/test_formulae.rb:
```diff
@@ -15,22 +15,25 @@ def initialize(tap:, git:, dry_run:, fail_fast:, verbose:)
-          event_path = ENV.fetch("GITHUB_EVENT_PATH")
-          event_payload = JSON.parse(File.read(event_path))
+        github_event_payload = JSON.parse(File.read(github_event_path))
+        return if github_event_payload.dig("pull_request", "head", "repo", "owner", "login") != "Homebrew"
```
```suggestion
        return if github_event_payload.dig("pull_request", "head", "repo", "owner", "login") != ENV.fetch("GITHUB_REPOSITORY_OWNER")
```

Maybe?

Will allow it to work in third-party taps.

carlocab replied:
Thanks, applied.

carlocab commented on lib/test_formulae.rb:
```diff
@@ -15,22 +15,28 @@ def initialize(tap:, git:, dry_run:, fail_fast:, verbose:)
+        github_event_payload = JSON.parse(File.read(github_event_path))
+        head_repo_owner = github_event_payload.dig("pull_request", "head", "repo", "owner", "login")
+        head_from_fork = head_repo_owner != ENV.fetch("GITHUB_REPOSITORY_OWNER")
+        maintainer_fork = JSON.parse(HOMEBREW_MAINTAINER_JSON.read).include?(head_repo_owner)
+        return if head_from_fork && !maintainer_fork
```
This will need Homebrew/brew#15491.

MikeMcQuaid commented:
> It may work if that step is done as a pull_request_target, separate from everything else.

This will entirely work to avoid workflow tampering.

carlocab commented on lib/test_formulae.rb:
```diff
@@ -97,33 +105,51 @@ def download_artifact_from_previous_run!(name)
-        wanted_artifact = response.fetch("artifacts").find { |artifact| artifact.fetch("name") == name }
+        artifacts = response.fetch("artifacts")
+        wanted_artifact = artifacts.find { |artifact| artifact.fetch("name") == artifact_name }
+        # If we didn't find the artifact that we wanted, fall back to the `event_payload` artifact.
+        wanted_artifact ||= artifacts.find { |artifact| artifact.fetch("name") == "event_payload" }
```
This logic now needs adjusting to account for the fact that `event_payload` will now be from `triage.yml`.

Bo98 commented on lib/test_formulae.rb:
```diff
@@ -15,22 +15,30 @@ def initialize(tap:, git:, dry_run:, fail_fast:, verbose:)
+        github_event_payload = JSON.parse(File.read(github_event_path))
+        head_repo_owner = github_event_payload.dig("pull_request", "head", "repo", "owner", "login")
+        head_from_fork = head_repo_owner != ENV.fetch("GITHUB_REPOSITORY_OWNER")
+        maintainer_fork = JSON.parse(HOMEBREW_MAINTAINER_JSON.read).include?(head_repo_owner)
```
```suggestion
        maintainer_fork = tap.official? && JSON.parse(HOMEBREW_MAINTAINER_JSON.read).include?(head_repo_owner)
```

We probably shouldn't have elevated permissions on third-party taps.

carlocab commented:
Fix for CI failure: Homebrew/brew#15493

carlocab commented:
Ok, I think I'm pretty happy with where this is (but I'm still in the process of testing it a bit).

Significant changes since previous discussions:
1. We now use `tap_git_revision` in the bottle JSON as the point of comparison for `git diff` to decide whether a formula has changed. (This is typically a child/successor commit of the `before` SHA1.)
2. Because we use the bottle JSON here, we also now track unexpected changes to local bottle JSON files in `verify_local_bottles`.
3. We no longer `exit 1` if `verify_local_bottles` errors, because this prevents stacktraces from being printed (given that we do `verify_local_bottles` inside an `ensure`). `test-bot` will still exit with an error if `verify_local_bottles` encounters a problem.

carlocab commented:
Will merge this tomorrow when I have time to monitor Homebrew/core.

Also happy to revert parts of this or fix forward if I end up merging this before you've had the chance to review properly.
